pkg/agent/nodeportlocal/k8s/npl_controller.go

```diff
@@ -420,33 +445,64 @@ func (c *NPLController) handleAddUpdatePod(key string, obj interface{}) error {
 
 	nplAnnotationsRequired := []NPLAnnotation{}
 
-	// first, check which rules are needed based on the Pod specification (ignoring NPL
-	// annotations) and make sure they are present. As we do so, we build the expected list of
-	// NPL annotations for the Pod.
+	hostPorts := make(map[string]int)
 	for _, container := range podContainers {
 		for _, cport := range container.Ports {
-			port := int(cport.ContainerPort)
-			podPorts[port] = struct{}{}
-			portData := c.portTable.GetEntryByPodIPPort(podIP, int(cport.ContainerPort))
-			if portData == nil { // rule does not exist
-				if int(cport.HostPort) > 0 {
-					klog.V(4).Infof("Host Port is defined for Container %s in Pod %s, thus extra NPL port is not allocated", container.Name, key)
-					nodePort = int(cport.HostPort)
-				} else {
-					nodePort, err = c.portTable.AddRule(podIP, port)
-					if err != nil {
-						return fmt.Errorf("failed to add rule for Pod %s: %v", key, err)
-					}
-				}
+			portProtoInt := buildPortProto(fmt.Sprint(cport.ContainerPort), string(cport.Protocol))
+			if int(cport.HostPort) > 0 {
+				klog.V(4).Infof("Host Port is defined for Container %s in Pod %s, thus extra NPL port is not allocated", container.Name, key)
+				hostPorts[portProtoInt] = int(cport.HostPort)
+			}
+			if cport.Name == "" {
+				continue
+			}
+			portProtoStr := buildPortProto(cport.Name, string(cport.Protocol))
+			if targetPortsStr.Has(portProtoStr) {
+				targetPortsInt.Insert(portProtoInt)
+			}
+		}
+	}
+
+	// targetPortsInt contains list of all ports that needs to be exposed for the Pod, including container ports
+	// for named ports present in targetPortsStr. If it is empty, then all existing rules and annotations for the
+	// Pod have to be cleaned up.
```

Comment:
I think we should add: "If a Service uses a named target port that doesn't match any named container port for the current Pod, no corresponding entry will be added to the targetPortsInt set by the code above."

Reply:
Updated